File: 1519. Number of Nodes in the Sub-Tree With the Same Label/dfs.py
class Solution:
    def countSubTrees(self, n: int, edges: List[List[int]], labels: str) -> List[int]:

        # build graph based on edges (adjacency list represntation)
        graph = defaultdict(list)

        for parent, child in edges:
            graph[parent].append(child)
            graph[child].append(parent)

        res = [0] * n

        def dfs(node, parent):

            count = Counter()

            nonlocal res

            for neighbor in graph[node]:
                if parent != neighbor:
                    # continue count descendant
                    count += dfs(neighbor, node)

            # if neighbor's alphabet equal to parent's alphabet
            count[labels[node]] += 1
            # accumulate the number of count
            res[node] = count[labels[node]]
            # return the collection for calculating the number of count
            return count
            
        # first traversal the root by a non-exist node called "-1" 
        dfs(0, -1)
        return res
        

File: 1519. Number of Nodes in the Sub-Tree With the Same Label/dfs.cpp
class Solution {
    vector<int> arr;
public:
    vector<int> countSubTrees(int n, vector<vector<int>>& edges, string labels) {
        arr.resize(n);

        vector<vector<int>> adj(n);
        for (auto edge: edges){
            adj[edge[0]].push_back(edge[1]);
            adj[edge[1]].push_back(edge[0]);
        }

        dfs(-1, 0, adj, labels);

        return arr;

    }
private:
    vector<int> dfs(int parent, int child, vector<vector<int>>& edges, string& labels) {
        vector<int> res(26);
        for (auto edge:edges[child]) {
            if (parent != edge) {
                vector<int> count = dfs(child, edge, edges, labels);

                for (int alph = 0; alph < 26; alph++)
                    res[alph] += count[alph];
            }
        }

        arr[child] = ++res[labels[child] - 'a'];
        return res;
    }
};